### PR TITLE
Add native user authorization, login, and clan management 

### PR DESCRIPTION
**Pull request by copperwater**
as well as a couple other minor updates and additions, but this is the main thing.

### CODE REVIEW

File: scoreboard/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
-# Generated by Django 3.2.7 on 2021-10-08 14:55
```
Comment (aoeixsz4) [inline scoreboard/migrations/0001_initial.py:1]: Should we remove this and stop tracking anything in migrations or, if we're deliberately keeping this what was the reason?

File: tnnt/views.py [hunk 2]
```diff
@@ -1,4 +1,18 @@
 from django.views.generic import TemplateView
+from django.views import View
+from django.shortcuts import render
+from scoreboard.models import Player, Clan
+from tnnt.forms import CreateClanForm, InviteMemberForm
+from django.http import HttpResponseRedirect
+from . import hardfought_utils # find_player
+
+def impossible(*args):
+    # Yes, I made impossible(). I'm too used to having it around.
```
Comment (aoeixsz4) [inline tnnt/views.py]: I like having this impossible() but would suggest maybe having this additionally call die/exit or whatever, so that it behaves like a panic.
Also, there in the comment you say it's useful for seeing where something “really is known to be impossible” which I would reword as something like “catching things that *should* be impossible but aren't.”

File: tnnt/templates/clanmgmt.html [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    <form method="post">
+        {% csrf_token %}
+        {{ create_clan_form }}
+        <button name="create_clan" type="submit">Create</button>
+    </form>
+  {% else %}
+    <h3>{{clan.name}}</h3>
+    <form method="post" class="inline-form">
+      {% csrf_token %}
+        <button name="leave" type="submit" {% if members.count == 1 %}disabled{% endif %}>
```
Comment (aoeixsz4) [inline tnnt/templates/clanmgmt.html:49]: Wouldn't we want to remove the clan if the last player left?

File: tnnt/views.py (hunk 2 above)
Comment (copperwater) [reply tnnt/views.py]: I don't think we should crash the server and bring it down for *everyone* when someone manages to trigger one of these impossible cases (some of which can intentionally be done with a malicious POST request).

File: tnnt/templates/clanmgmt.html (hunk 3 above)
Comment (copperwater) [reply tnnt/templates/clanmgmt.html:49]: Submitting a "leave" POST request when the player is the sole clan member will indeed disband the clan (it just calls the disband_clan function), but in the UI there's a big "Disband Clan" button anyway, so there's no reason to confusingly have "Leave" up there as well.

File: scoreboard/migrations/0001_initial.py (hunk 1 above)
Comment (copperwater) [reply scoreboard/migrations/0001_initial.py:1]: I am fine with gitignoring migrations

File: tnnt/views.py [hunk 4]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+        # if we got here, we're good to attempt the invite
+        try:
+            # retrieve player if exists in our database, if not
+            # but they are in dgl database, create the player in our
+            # database (we need to do this because we need to record
+            # them as having an invite)
+            invitee = hardfought_utils.find_player(request.POST['invitee'])
+            invitee.invites.add(player.clan)
+        except Player.DoesNotExist:
+            ctx['errmsg'] = 'No such player "' + request.POST['invitee'] + '" exists.'
```
Comment (aoeixsz4) [inline tnnt/views.py]: Not sure if they can do any actual damage with this, but I don't like unvalidated input being concatenated into the error message like this.

File: tnnt/views.py [hunk 5]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+
+        elif 'kick' in request.POST:
+            self.kick_member(request, player, ctx)
+
+        return render(request, self.template_name, self.get_context_data(**ctx))
+
+    # Helper function triggered when a create_clan POST request comes in
+    def create_clan(self, request, player, ctx):
+        create_clan_form = CreateClanForm(request.POST)
+        new_clan_name = request.POST['clan_name']
```
Comment (aoeixsz4) [inline tnnt/views.py]: This will error if a user POSTs with 'create_clan' but no 'clan_name' in the request.

File: tnnt/views.py [hunk 6]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+            ctx['errmsg'] = 'Only admins can do this'
+            return None
+
+        return otherplayer
+
+    # Helper function triggered when "Rescind" is clicked on an invited player
+    def rescind_invite(self, request, player, ctx):
+        # TODO? all of these helpers should test that the key is in POST, and if
+        # not, log an error and return
+        rescindee_id = request.POST['rescind_id']
```
Comment (aoeixsz4) [inline tnnt/views.py:349]: Will error on invalid POST data from funny hacker user.

File: tnnt/views.py [hunk 7]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+        rescindee = self.clan_admin_other_member_checks(request, player, ctx, rescindee_id)
+        if rescindee is None:
+            return
+
+        # if we passed checks, we're good to rescind the invite
+        rescindee.invites.remove(player.clan)
+
+    # Helper function triggered when "Make Admin" is clicked on a clan member
+    def make_admin(self, request, player, ctx):
+        new_admin_id = request.POST['kick_or_admin_id']
```
Comment (aoeixsz4) [inline tnnt/views.py:362]: Will error on bad POST data.

File: tnnt/views.py [hunk 8]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+            ctx['errmsg'] = '%s is already a clan admin!' % (new_admin.name)
+            return
+
+        # if we passed checks, we're good to make this person an admin
+        new_admin.clan_admin = True
+        new_admin.save()
+
+    # Helper function triggered when "Kick" is clicked on a clan member
+    def kick_member(self, request, player, ctx):
+        kickee_id = request.POST['kick_or_admin_id']
```
Comment (aoeixsz4) [inline tnnt/views.py:390]: Will error on bad POST data.

File: tnnt/views.py [hunk 9]
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
...
+        kickee = self.clan_admin_other_member_checks(request, player, ctx, kickee_id)
+        if kickee is None:
+            return
+
+        # other player must be in this clan
+        if player.clan != kickee.clan:
```
Comment (aoeixsz4) [inline tnnt/views.py:397]: I would use identity rather than equality here.

File: tnnt/views.py (hunk 2 above)
Comment (aoeixsz4) [reply tnnt/views.py]: I assume it uses some async tasks and multithreading under the hood and only will crash the page, but I have yet to test this for sure.
In fact, at least in development-mode errors in code get logged to the terminal output and a detailed error page is displayed in the browser. When running in production however, the user won't be served this detailed stuff. Server doesn't totally crash tho. Then again deliberately calling die() or exit() maybe WOULD kill the server. Instead, I suggest we throw an exception.

File: tnnt/templates/clanmgmt.html (hunk 3 above)
Comment (aoeixsz4) [reply tnnt/templates/clanmgmt.html:49]: Ahhhh okay, I gotchu.

File: tnnt/views.py [hunk 10]
```diff
@@ -11,3 +25,305 @@ class AboutView(TemplateView):
...
+
+        # can't join a clan if already in one
+        if player.clan:
+            ctx['errmsg'] = "You can't join that clan because you're already in one"
+            return
+
+        # clan id actually needs to exist
+        # this could theoretically happen normally, where one player loads the
+        # clan management page and has an invite for another clan, which has
+        # disbanded by the time they submit the join request.
```
Comment (deepy) [inline tnnt/views.py:276]: shouldn't the delete for the clan cascade and delete the invites as well?

Comment (copperwater) [reply tnnt/views.py:276]: Should happen automatically by django's managing of ManyToManyField relationships. (Confirmed by querying the database directly what happens if I create a clan, invite a bunch of people, and then disband it -- the scoreboard_player_invites table is left the same as it was before I made the clan.)

File: tnnt/views.py (hunk 9 above)
Comment (copperwater) [reply tnnt/views.py:397]: Using `if player.clan is not kickee.clan:` results in the admin not being able to kick anyone.

File: tnnt/views.py (hunk 5 above)
Comment (deepy) [reply tnnt/views.py]: shouldn't this use `create_clan_form.cleaned_data[]` after checking that the form `.is_valid()`?

File: tnnt/views.py (hunk 10 above)
Comment (deepy) [reply tnnt/views.py:276]: You'd have to specify `on_delete` but having deletes cascade is generally the preferred solution

File: tnnt/views.py (hunk 9 above)
Comment (aoeixsz4) [reply tnnt/views.py:397]: oh weird, nvm then

File: tnnt/views.py (hunk 5 above)
Comment (aoeixsz4) [reply tnnt/views.py]: not sure lol

Comment (copperwater) [reply tnnt/views.py]: done. perhaps @deepy can recommend how to make all the other forms on the page proper ones in forms.py, and make them validate by default as well, as that's something I just didn't have time for.

File: tnnt/hardfought_utils.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    except Player.DoesNotExist as e:
+        try:
+            dgl_curs = get_dgl_cursor()
+            uname = dgl_curs.execute('SELECT username FROM dglusers WHERE username = ?',
+                                     (findname,)).fetchone()
+        except sqlite3.Error as sqlite_e:
+            print('ERROR connecting to or executing SQL on sqlite database')
+            print('  db path:', DGL_DATABASE_PATH)
+            print('  username:', username)
+        if uname is None:
```
Comment (aoeixsz4) [inline tnnt/hardfought_utils.py:29]: This behaviour will mean I need a dummy dgl database for testing that has one each of everyone's nick...
I suppose I can generate such a thing in a similar way to your alice, bob, carol, dan one?

**Review by aoeixsz4 — APPROVED** [review]
ship it
